pages/exercises/[lessonSlug].tsx

```diff
@@ -104,55 +108,71 @@ const Exercises: React.FC<QueryDataProps<GetExercisesQuery>> = ({
 }
 
 type ExerciseData = {
+  id: number
   problem: string
   answer: string
   explanation: string
 }
 
 type ExerciseProps = {
-  exercise: ExerciseData
-  setExerciseIndex: React.Dispatch<React.SetStateAction<number>>
   lessonTitle: string
-  hasPrevious: boolean
-  hasNext: boolean
-  submitUserAnswer: (userAnswer: string) => void
+  exercises: ExerciseData[]
+  userAnswers: Record<number, string>
+  submitUserAnswer: (exerciseId: number, userAnswer: string) => void
+  onExit: (userAnswers: Record<number, string>) => void
 }
 
 const Exercise = ({
-  exercise,
-  setExerciseIndex,
   lessonTitle,
-  hasPrevious,
-  hasNext,
-  submitUserAnswer
+  exercises,
+  userAnswers,
+  submitUserAnswer,
+  onExit
 }: ExerciseProps) => {
   const [answerShown, setAnswerShown] = useState(false)
   const [message, setMessage] = useState(Message.EMPTY)
+  const [exerciseIndex, setExerciseIndex] = useState(0)
+  const [localUserAnswers, setLocalUserAnswers] = useState(userAnswers)
+  const exercise = exercises[exerciseIndex]
+
+  const hasPrevious = exerciseIndex > 0
+  const hasNext = exerciseIndex < exercises.length - 1
 
   return (
     <div className={`mx-auto ${styles.exercise__container}`}>
       <button
         className="btn ps-0 d-flex align-items-center"
-        onClick={() => setExerciseIndex(-1)}
+        onClick={() => onExit(localUserAnswers)}
       >
         <ArrowLeftIcon size="medium" aria-label="Exit" />
       </button>
 
       <h1 className="mb-4 fs-2">{lessonTitle}</h1>
       <ExerciseCard
+        key={exercise.id}
         problem={exercise.problem}
         answer={exercise.answer}
         explanation={exercise.explanation}
         answerShown={answerShown}
         setAnswerShown={setAnswerShown}
         message={message}
         setMessage={setMessage}
-        submitUserAnswer={submitUserAnswer}
+        submitUserAnswer={userAnswer => {
+          setLocalUserAnswers({
+            ...localUserAnswers,
+            [exercise.id]: userAnswer
+          })
+          submitUserAnswer(exercise.id, userAnswer)
+        }}
       />
       <div className="d-flex justify-content-between mt-4">
         {hasPrevious ? (
           <button
-            onClick={() => setExerciseIndex(i => i - 1)}
+            onClick={() => {
+              setExerciseIndex(i => i - 1)
+              setAnswerShown(false)
+              setMessage(Message.EMPTY)
```

Comment:
The reason why we now have to set the message to EMPTY and set the answer to be not shown is because we are no longer completely rerendering the `Exercise` component (it never gets completely rerendered now because we removed `key` prop and because we don't update any parent state until after we exit the `Exercise` component).